pkg/urbit/noun/retrieve.c

```diff
@@ -1056,24 +1056,24 @@ u3r_met(c3_y    a_y,
       case 0:
       case 1:
       case 2: {
-        /* col_w: number of bits in (daz_w)
-        ** bif_w: number of bits in (b)
-        */
-        c3_w bif_w, col_w;
+        //  number of bits in [b], rounded up
+        //
+        c3_w bif_w = (gal_w << 5) + c3_bits_word(daz_w) + ((1 << a_y) - 1);
 
-        col_w = c3_bits_word(daz_w);
-        bif_w = col_w + (gal_w << 5);
+        if ( bif_w < gal_w ) {
```

Comment:
Can you explain the motivation behind using `gal_w` as the sentinel for overflow? It strikes me as imprecise, because if I'm following correctly, we should be able to specify the exact maximum value of `gal_w` to avoid overflow.

The largest value that `gal_w` can be without overflow is (call it `x`):
```c
0b0000_0011_1111_1111_1111_1111_1111_1111
```
So `x << 5` is:
```c
0b0111_1111_1111_1111_1111_1111_1110_0000
```
Adding in `32`, which is the greatest number of possible bits in `daz_w`:
```c
0b1000_0000_0000_0000_0000_0000_0000_0000
```
In other words, why aren't we checking that `gal_w` is less than or equal to `0b0000_0011_1111_1111_1111_1111_1111_1111` to ensure that overflow will not occur?

Reply:
This is a good point, but your numbers aren't quite maximal:

```
> `@ux`(rsh [0 5] (sub (dec (bex 32)) 32))
0x7ff.fffe
> `@ub`(rsh [0 5] (sub (dec (bex 32)) 32))
0b111.1111.1111.1111.1111.1111.1110
```

Or in C

```
(lldb) p/x (((1ULL << 32) - 1) - 32) >> 5
(unsigned long long) $2 = 0x0000000007fffffe
```

Where `(1ULL << 32) - 1` is just `UINT32_MAX`.

Reply:
Actually, this isn't right either. We're rounding up, so the maximum addend isn't `32`, it's `35` (`32 + (1 << 2) - 1`).

Reply:
I'd normally use the constant here with a comment, but both gcc and clang constant-fold it, even at `-O0`.